client/subscription-edit-page/index.tsx

```diff
@@ -0,0 +1,285 @@
+ *
+ * This could be a shorter method, but because the customer select
+ * element uses select2, it does not emit the typical `change` event.
+ *
+ * @param {(userId: number) => void} callback The callback to call when the customer is changed.
+ * @return {() => void} The cleanup function.
+ */
+export const addCustomerSelectListener = (
+	cache: CachedUserData,
+	ajaxUrl: string,
+	nonce: string,
+	setUser: ( userId: number ) => void,
+	setCache: ( cache: CachedUserData ) => void
+): ( () => void ) => {
+	const element = document.getElementById( 'customer_user' );
+	const customerUserSelect =
+		element instanceof HTMLSelectElement ? element : null;
+
+	if ( ! customerUserSelect ) {
+		return (): void => {
+			// No-op cleanup function when an element is not found.
+		};
+	}
+
+	// Wrap in an internal callback to load the select's value and tokens.
+	const internalCallback = async () => {
+		const newUserId = parseInt( customerUserSelect.value, 10 ) || 0;
+
+		setUser( newUserId );
+
+		// Loaded, loading, or errored out, we do not need to load anything.
+		if ( hasEntry( cache, newUserId ) ) {
+			return;
+		}
+
+		const cacheWithLoading = startLoading( cache, newUserId );
+		setCache( cacheWithLoading );
+
+		try {
+			const response = await fetchUserTokens( newUserId, ajaxUrl, nonce );
+			if ( undefined === response ) {
+				throw new Error(
+					__(
+						'Failed to fetch user tokens. Please reload the page and try again.',
+						'woocommerce-payments'
+					)
+				);
+			}
+
+			setCache(
+				tokensLoaded( cacheWithLoading, newUserId, response.tokens )
+			);
+		} catch ( error ) {
+			setCache(
+				loadingFailed(
+					cacheWithLoading,
+					newUserId,
+					error instanceof Error
+						? error.message
+						: __( 'Unknown error', 'woocommerce-payments' )
+				)
+			);
+		}
+	};
+
+	// Add the listener with the right technique, as select2 does not emit <select> events.
+	jQuery( customerUserSelect ).on( 'select2:select', internalCallback );
+	customerUserSelect.addEventListener( 'change', internalCallback );
+
+	return () => {
+		jQuery( customerUserSelect ).off( 'select2:select', internalCallback );
+		customerUserSelect.removeEventListener( 'change', internalCallback );
+	};
+};
+
+/**
+ * Renders a payment method select or loading indicator.
+ */
+export const PaymentMethodSelect = ( {
+	inputName,
+	initialValue,
+	initialUserId,
+	initialCache,
+	nonce,
+	ajaxUrl,
+}: PaymentMethodSelectProps ) => {
+	const [ value, setValue ] = useState< number >( initialValue );
+	const [ userId, setUserId ] = useState< number >( initialUserId );
+	const [ cache, setCache ] = useState< CachedUserData >( initialCache );
+
+	// If there is no value but a user, try to fall back to the default token.
+	useEffect( () => {
+		if ( 0 !== userId && value === 0 ) {
+			const defaultTokenId = getDefaultTokenId( cache, userId );
+			if ( value !== defaultTokenId ) {
+				setValue( value );
+			}
+		}
+	}, [ userId, value, cache ] );
```

Comment:
```suggestion
	// If there is no value but a user, try to fall back to the default token.
	useEffect( () => {
		if ( 0 !== userId && value === 0 ) {
			const defaultTokenId = getDefaultTokenId( cache, userId );
			if ( value !== defaultTokenId ) {
				setValue( defaultTokenId );
			}
		}
	}, [ userId, value, cache ] );
```

Given the description above the `useEffect`, should this have been the intent?


Also, is `cache` necessary as part of the dependencies?

Reply:
> Given the description above the `useEffect`, should this have been the intent?

Absolutely. I'm not sure how I let this through. Maybe there was some TS error in the test file, and I was actually looking at an outdated front-end, as this was one of the later changes. 🤔

I updated it to use `defaultTokenId`.

> Also, is cache necessary as part of the dependencies?

Once tokens finish loading for a new user, we need to use the cache update to trigger the effect. `userId` and `value` are not directly changed when tokens get loaded.